deeptime/markov/msm/tram/_tram_dataset.py

```diff
@@ -0,0 +1,494 @@
+                (i_s, j_s) = connectivity_fn(self.ttrajs, self.dtrajs, self.bias_matrices, all_state_counts,
+                                             self.n_therm_states, self.n_markov_states, connectivity_factor,
+                                             callback)
+            print((i_s, j_s))
+            # add transitions that occurred within each thermodynamic state. These are simply the connected sets:
+            for k in range(self.n_therm_states):
+                for cset in estimator.fit_fetch(self.dtrajs[k]).connected_sets():
+                    i_s.extend(list(cset[0:-1] + k * self.n_markov_states))
+                    j_s.extend(list(cset[1:] + k * self.n_markov_states))
+
+            # turn the list of transitions into a boolean matrix that has a one whenever a transition has occurred
+            data = np.ones(len(i_s), dtype=np.int32)
+            dim = self.n_therm_states * self.n_markov_states
+            sparse_transition_counts = sp.sparse.coo_matrix((data, (i_s, j_s)), shape=(dim, dim))
+
+            # Now we have all possible paths in the list of transitions. Get the connected set of that
+            overlap_counts_model = TransitionCountModel(sparse_transition_counts)
+            connected_states_ravelled = overlap_counts_model.submodel_largest(directed=False).state_symbols
+
+            # unravel the index and combine all separate csets to one cset
+            connected_states = np.unravel_index(connected_states_ravelled, (self.n_therm_states, self.n_markov_states),
+                                                order='C')
+
+            return full_counts_model.submodel(np.unique(connected_states[1]))
+
+    def _find_trajectory_fragments(self):
+        """ Find all trajectory fragments given the discrete trajectories and the thermodynamic state indices of each
+        sample. Get rid of any negative state indices.
+        If no replica exchange was done, this will simply return all positive values from dtrajs. If ttrajs are
+        given, we assume replica exchange was done and restructure dtrajs in such a way that every replica-exchange
+        swap is the start of a new trajectory.
+
+        Returns
+        -------
+        dtraj_fragments : list(List(int))
+           A list that contains for each thermodynamic state the fragments from all trajectories that were sampled at
+           that thermodynamic state. fragment_indices[k][i] defines the i-th fragment sampled at thermodynamic state k.
+           The fragments are be restricted to the largest connected set and negative state indices are excluded.
+        """
+        if self.ttrajs is None or len(self.ttrajs) == 0:
+            # No ttrajs were given. We assume each trajectory in dtrajs was sampled in a distinct thermodynamic state.
+            # The thermodynamic state index equals the trajectory index, and the dtrajs are unchanged.
+            return [[self.dtrajs[k][self.dtrajs[k] >= 0]] for k in range(self.n_therm_states)]
```

Comment:
you check here for negative entries in dtrajs, can this lead to artifacts if they're just sliced out? like `[0, 0, 2, -1, 2, 1, 0]` leads to `[0, 0, 2, 2, 1, 0]` instead of maybe `[0, 0, 2], [2, 1, 0]` (transition 2 -> 2 doesn't exist)

Reply:
Will be resolved in follow-up PR